src/mito/src/table.rs

```diff
@@ -127,35 +142,60 @@ impl<R: Region> Table for MitoTable<R> {
         _limit: Option<usize>,
     ) -> TableResult<PhysicalPlanRef> {
         let read_ctx = ReadContext::default();
-        let snapshot = self
-            .region
-            .snapshot(&read_ctx)
-            .map_err(BoxedError::new)
-            .context(table_error::TableOperationSnafu)?;
-
-        let projection = self
-            .transform_projection(&self.region, projection.cloned())
-            .map_err(BoxedError::new)
-            .context(table_error::TableOperationSnafu)?;
-        let filters = filters.into();
-        let scan_request = ScanRequest {
-            projection,
-            filters,
-            ..Default::default()
-        };
-        let mut reader = snapshot
-            .scan(&read_ctx, scan_request)
-            .await
-            .map_err(BoxedError::new)
-            .context(table_error::TableOperationSnafu)?
-            .reader;
-
-        let schema = reader.schema().clone();
-        let stream_schema = schema.clone();
+        let mut readers = Vec::with_capacity(self.regions.len());
+        let mut first_schema: Option<Arc<Schema>> = None;
+
+        // TODO(hl): Currently the API between frontend and datanode is under refactoring in
+        // https://github.com/GreptimeTeam/greptimedb/issues/597 . Once it's finished, query plan
+        // can carry filtered region info to avoid scanning all regions on datanode.
+        for region in self.regions.values() {
+            let snapshot = region
+                .snapshot(&read_ctx)
+                .map_err(BoxedError::new)
+                .context(table_error::TableOperationSnafu)?;
+            let projection = self
+                .transform_projection(region, projection.cloned())
+                .map_err(BoxedError::new)
+                .context(table_error::TableOperationSnafu)?;
+            let filters = filters.into();
+            let scan_request = ScanRequest {
+                projection,
+                filters,
+                ..Default::default()
+            };
+            let reader = snapshot
+                .scan(&read_ctx, scan_request)
+                .await
+                .map_err(BoxedError::new)
+                .context(table_error::TableOperationSnafu)?
+                .reader;
+
+            let schema = reader.schema().clone();
+            if let Some(first_schema) = &first_schema {
+                ensure!(
+                    first_schema.version() == schema.version(),
```

Comment:
We may need to add a TODO here, the schemas in different regions may be different when altering table schema in future.